File: scanner/actions/DLNAAction.js
var exec = require('child_process').exec
const MediaURLAction = require('./MediaURLAction')
const pythonExtensionsPath =
  __dirname + '/../../.virtualenv/bin/python ' + __dirname + '/../../python_extensions'

class DLNAAction extends MediaURLAction {
  process() {
    const envVars = this.envVars()
    // Stop playback here for better UX
    const scriptPathCC =
      pythonExtensionsPath + '/chromecast.py --chromecast_ip ' + this.config.chromecast_ip + ' stop'
    exec(scriptPathCC, {env: envVars}, function(error, stdout, stderr) {
      console.log(stdout, stderr, error)
    })
    // Fetch all DLNA titles, and use the mappings to get a media URL
    const scriptPathDLNA =
      pythonExtensionsPath + '/dlna.py --dlnaserver_ip ' + this.config.dlnaserver_ip
    exec(scriptPathDLNA, (error, stdout, stderr) => {
      if (error || stderr) {
        console.log(error)
      } else {
        const dlnaMappings = JSON.parse(stdout)
        const random = this.card.uri.split(':')[0]
        const regex = this.card.uri.split(':')[1]
        console.log('regex', regex)
        // Allow using regex for getting multiple urls
        this.card.uri =
          random +
          (random != '' ? ':' : '') +
          Object.entries(dlnaMappings)
            .filter(val => val[0].match(regex) !== null)
            .map(val => val[1])
            .join(' ')
        // Run as a normal media URL
        super.process()
      }
    })
  }
}

module.exports = DLNAAction

File: scanner/actions/dlna_action.py
import re

from .mediaurl_action import MediaUrlAction
from utils.dlna import parse_dlna


class DLNAAction(MediaUrlAction):
    def process(self):
        self.chromecast.stop()
        uri = self.card['uri']
        dlna_mappings = parse_dlna(self.config['dlnaserver_ip'])
        flag = uri.split(':')[0]
        regex = uri.split(':')[1]
        new_uri = ''
        if flag != '':
            new_uri = '{}:'.format(flag)
        for title, url in dlna_mappings.items():
            if re.search(regex, title) is not None:
                new_uri += "{} ".format(url)
        self.card['uri'] = new_uri
        super().process()
